### DISCUSSION TITLE
use of own certificates and https-pass-through

### DISCUSSION THREAD

**Discussion by geigervibe** [NginxProxyManager/nginx-proxy-manager#1598] [Q&A]
Hi all!

I'm currently use traefik but my configuration will not work, for that I wonder if NPM is a better option and I wonder, if NPM supports the use of own certificates and https-pass-through. For both I could not find anything in the documentation (https://nginxproxymanager.com/setup/). If it supports these functions, were can I find more to read about, how to setup?

I need at least two routings:
https > npm > http/container1
https > npm > https/container2 (passthrough)
And as told using own certificate and NOT letsencrypt...

Thanks and regards...

**Comment by chaptergy** [comment]
It is currently not possible, I made a PR which would add this functionality but I'm not sure it will get merged, as it is a huge change and comes with a load of issues. See https://github.com/jc21/nginx-proxy-manager/issues/853

**Comment by geigervibe** [comment]
I was able to answer one question myself, my own certificates can be integrated via the web GUI, as 'Custom' ones. Nothing special needs to be set up in the composer.

Thanks for the clarification, @chaptergy . If I have understood correctly, it is not possible to use entry port 443 for both proxy and ssl-passthrough at the same time, regardless of which edition of nginx is used? No possibility at all?

If I interpret the example-configuration under https://admintuts.net/server-admin/docker/nginx-reverse-proxy-with-ssl-pass-through-load-balancing/ (see 'Bonus: Configure Nginx Reverse Proxy SSL Passthrough SNI For Docker') correctly, the port 3443 is used for the ssl-passthrough and port 443 for proxy? Did I interpreted that correctly?

**Reply by chaptergy** [reply]
SSL passthrough is possible in theory, and it is implemented in the PR I linked, but it comes with a whole lot of caveats, issues and performance penalties, so I'm not sure it will ever be merged. In order to make ssl passthorugh work, everything but the non-ssl-passthrough hosts would have to be proxied twice.

**Reply by geigervibe** [reply]
Yes, I understood that based on your linked post. But still I'm not sure, if one can use in general one and the same entry port (ex. 443) for both proxy and passthrough no matter what edition of nginx is used. Because, if not, that would be a knock out in the first place and by edition I mean for example using a full installation of nginx and config the files from hand, like the config I linked in my last post. This config looks like that it use two entry ports one for the proxied trafic and one for the passthrough stream. Did I interpreted that config correctly?

**Reply by chaptergy** [reply]
I can't really follow, but npm is just generating config files for nginx, the nginx used is just normal nginx. The config you linked is using different ports, however it is using port 80 (http port) for the normal http hosts to forward to the http version of the app, and the stream / passthrough hosts to forward to the https version of the app on port 443 (https port).

**Reply by geigervibe** [reply]
Sorry, my english is not that good and therefor confusing. I thought when I use native nginx I would have more possibilities than using npm. OK, the underlaying nginx for npm is a full installation of nginx, I did not know that. What I would want is the following:

Browser https://app1.example.com (443) > example.com/nginx/proxy > http/container1
Browser https://app2.example.com (443) > example.com/nginx/sslpassthrough > https/container2

Would that be possible when using a host with only one IP were normal nginx is running? I read at some places, that you would need at least two entry ports or two IP's for that. But I'm not sure, if I got that right. If that is really the case, then nginx would not be an option for this kind of configuration (besides the other caveats you mention).

**Reply by chaptergy** [reply]
You do have more options when you create the nginx config files yourself, npm limits the options it exposes to make it easy to use. Otherwise you would have to mount custom configs into the container using the docker-compose.  

So why do you want to proxy it twice? Why https://app1.example.com (443) > example.com/nginx/proxy > http/container1 instead of just doing https://app1.example.com (443) > http/container1?

Proxying multiple websites on the same port is exactly what a reverse proxy does. When managing the SSL certificate in NPM it is very easy and you can just add proxy hosts which do exactly that. If you want the app behind the proxy to handle the SSL certificate, it is much more complicated and not possible through the interface of the current npm version. But it is possible in nginx in general.

**Reply by geigervibe** [reply]
Thank you for taking the time to answer my further questions, but I think we have some misunderstandings here.

With my illustration I wanted to mark the following path: 
browser > proxy > endpoint 
.. so no double proxy chain.

Even if the question of which answer I am wrestling with has nothing to do with npm but rather with nginx in general, it will help me to focus the right way for the future.

As shown in the illustration in previous post, I continue to wonder whether it is possible that the same entry port 443 can be used for both proxy and ssl-passthrough at the same time via one nginx-instance. 

There are statements on the Internet that this is not possible, but I am not sure whether I have understood that correctly.

If that really doesn't work, I would have to deal with traefik again and neither hope/wait for a merge nor setup nginx by hand.

Up until now I have managed to access the app2 (see illustration) via npm and proxy scheme https, but that has nothing to do with ssl-passthrough, but this way I can currently reach both app1 and app2 via port 443 in my lab environment.

**Reply by chaptergy** [reply]
Am I now totally misunderstanding you? As I have said in my previous responses, having a ssl terminating proxy and a ssl passthrough proxy is possible on the same port, though much harder than just having ssl teminating proxies. Is that not what you want to know?

**Reply by geigervibe** [reply]
Yes, that's exactly what I wanted to know. Although I have now looked again where I have read that that is just not possible (using the same port). And I have now found one of the posts and it is actually on the same page where you have posted your PR:

[NGINX can do SSL passthru, but it can't do it on the same ports being serviced for normal NGINX services. You can do only one or the other.](https://github.com/jc21/nginx-proxy-manager/issues/853#issuecomment-785541267)

So now I'm even more confused. But I will leave it at that, instead of passthrough I now use the https scheme, although it is not the same, it serves the purpose for now.
